drivers/filesystems/fastfat/misc.c

```diff
@@ -336,19 +336,64 @@ static
 VOID
 NTAPI
 VfatDoRequest(
-    PVOID IrpContext)
+    PVOID Context)
 {
+    PVFAT_IRP_CONTEXT IrpContext = Context;
+    PDEVICE_EXTENSION DeviceExt;
+    KIRQL OldIrql;
+
     InterlockedDecrement(&QueueCount);
-    DPRINT("VfatDoRequest(IrpContext %p), MajorFunction %x, %d\n",
-           IrpContext, ((PVFAT_IRP_CONTEXT)IrpContext)->MajorFunction, QueueCount);
-    VfatDispatchRequest((PVFAT_IRP_CONTEXT)IrpContext);
+
+    if (IrpContext->Stack->FileObject != NULL)
+    {
+        DeviceExt = IrpContext->Stack->DeviceObject->DeviceExtension;
+        ObReferenceObject(DeviceExt->VolumeDevice);
+    }
+
+    do
+    {
+        DPRINT("VfatDoRequest(IrpContext %p), MajorFunction %x, %d\n",
+               IrpContext, IrpContext->MajorFunction, QueueCount);
+        VfatDispatchRequest(IrpContext);
+        IrpContext = NULL;
+
+        /* Now process any overflow items */
```

Comment:
If I understand correctly the code, it's processing just one item at a time?
I was wondering why you are not doing something like this instead:
```c
while (!IsListEmpty(&DeviceExt->OverflowQueue))
{
    IrpContext = CONTAINING_RECORD(RemoveHeadList(&DeviceExt->OverflowQueue),
                                   VFAT_IRP_CONTEXT,
                                   WorkQueueItem.List);

    // Proceed as in the rest...
}
```

Reply:
The first item is the one passed to the WorkItem via the context. After that it processes items from the overflow queue (which may not exist, if there is no FileObject).
I could separate the first item from the loop if you feel strongly that that would be clearer, e.g.
```
VfatDispatchRequest();

if (DeviceExt)
{
    while (!IsListEmpty())
    {
        VfatDispatchRequest();
    }
}
```